workbench/x3/rule_defns/gen_benchmark_gram.cpp

```diff
@@ -0,0 +1,223 @@
+      gfi()<<"#if RULE2RHS_WHICH == RULE2RHS_CTX_MI_REC\n";
+      ++gfi;
+          gfi()<<"auto const& ctx=make_defs\n";
+          gfi()<<"( "<<d_(num_rules-1)<<"\n";
+          gfi()<<");"<<"\n";
+      --gfi;
+      gfi()<<"#else\n";
+        ++gfi;
+        gfi()<<"auto const& ctx=empty_context();\n";
+        gfi()<<"#if RULE2RHS_WHICH == RULE2RHS_GET_RHS_REC\n";
+          ++gfi;
+          gfi()<<"BOOST_SPIRIT_DEFINE\n";
+          gfi()<<"( "<<d_(num_rules-1)<<"\n";
+          gfi()<<");"<<"\n";
+          --gfi;
+        gfi()<<"#endif\n";
+        --gfi;
+      gfi()<<"#endif //RULE2RHS_WHICH == RULE2RHS_CTX_MI_REC\n";
+    --gfi;
+    gfi()<<"#elif RULE2RHS_WHICH != RULE2RHS_GET_RHS_CRTP\n";
+    ++gfi;
+      for(i_rule=0;i_rule<num_rules; ++i_rule)
+        gfi()<<r_lhs<def>(i_rule)+"="+rhs<rul>(i_rule,num_rules)+";\n";
+      gfi()<<"auto const& start="<<d_(num_rules-1)<<";\n";
+      gfi()<<"#if RULE2RHS_WHICH == RULE2RHS_CTX_MI_ALL\n";
+      ++gfi;
+          gfi()<<"auto const& ctx=make_defs\n";
+          for(i_rule=num_rules-1; i_rule>=0; --i_rule)
+          {
+            if(i_rule==num_rules-1)
+              gfi()<<"( ";
+            else
+              gfi()<<", ";  
+            gfi.os()<<d_(i_rule)<<"\n";
+          }
+          gfi()<<");"<<"\n";
+      --gfi;
+      gfi()<<"#else\n";
+        ++gfi;
+        gfi()<<"auto const& ctx=empty_context();\n";
+        gfi()<<"#if RULE2RHS_WHICH == RULE2RHS_GET_RHS_ALL\n";
+          ++gfi;
+          gfi()<<"BOOST_SPIRIT_DEFINE\n";
+          for(i_rule=num_rules-1; i_rule>=0; --i_rule)
+          {
+            if(i_rule==num_rules-1)
+              gfi()<<"( ";
+            else
+              gfi()<<", ";  
+            gfi.os()<<d_(i_rule)<<"\n";
+          }
+          gfi()<<");"<<"\n";
+          --gfi;
+        gfi()<<"#endif\n";
+        --gfi;
+      gfi()<<"#endif //RULE2RHS_WHICH == RULE2RHS_CTX_MI_ALL\n";
+    --gfi;
+    gfi()<<"#else//RULE2RHS_GET_RHS_CRTP\n";
+    ++gfi;
+      gfi()<<"struct gram_deriv: gram_base<gram_deriv>\n";
+      gfi()<<"{\n";
+      ++gfi;
+        rule_decls( gfi, num_rules);
+        gfi()<<"BOOST_SPIRIT_GET_RHS\n";
+        ++gfi;
+          gfi()<<"(\n";
+          ++gfi;
+          for(i_rule=0; i_rule<num_rules; ++i_rule)
+          {
+            if(i_rule>0) 
+            {
+              --gfi; 
+              gfi()<<",\n";
+              ++gfi;
+            }
+            gfi()<<"(\n";
+            ++gfi;
+              gfi()<<r_lhs<rul>(i_rule)<<"="<<rhs<rul>(i_rule,num_rules)<<"\n";
+            --gfi;
+            gfi()<<")\n";
+          }
+          --gfi;
+          gfi()<<")\n";
+        --gfi;
+      --gfi;
+      gfi()<<"};\n";
+      gfi()<<"gram_deriv g;\n";
+      gfi()<<"auto const& start=g.get_rhs(id<"<<num_rules-1<<">{});\n";
+      gfi()<<"auto const& ctx = empty_context();\n";
+    --gfi;
+    gfi()<<"#endif//end:USE_RHS_DEF\n";
+  }
+  gf.close();
+  return 0;
+}
```

Comment:
Code format of this makes it nearly impossible to understand :(

Reply:
I assume you're referring to the code indentation.  I added
comments at the top of the code to explain that the odd
indentation is to help debugging.  Basically, the code
indentation reflects the output indentation to make
debugging the code much easier.  This become
really obvious once you try figuring out where
to put the ++gfi's and --gfi's while debugging the code.

Reply:
I am mostly about first hundred lines (this actually problem of all the PR).

_Suggestion: you really should use RAII instead of `++gfi's and --gfi's`._

Reply:
The rationale for that is that the function definitions form a dictionary, and the convention for dictionaries is the what's being
defined is aligned on the far left.  Thus, the function name is on the
far left.  Of course this is unconventional, but conventional is not always logical.  However, I will change the code conform the code to
convention by putting both the return type, the function name, and the
argument list on 1 line, if they fit.  If they don't fit, I'd break
*after* the comma for the argument closest to column 70 or thereabouts.

Reply:
Sorry, I don't understand how RAII applies here.  The ++gfi and --gfi change the current indentation of the current output file.  RAII is
about initializing a resource at the time of acquisition.  Well gfi
is acquired at the beginning of the block and destructed at the end of
that block but the indentation and outputs must happen in between; so,
how can RAII be used to do this?